### DISCUSSION TITLE
Iterable dynamic imports

### DISCUSSION THREAD

**Discussion by NuroDev** [vercel/next.js#35351] [Ideas]
### What
A nice improvement to the [dynamic import](https://nextjs.org/docs/advanced-features/dynamic-import) feature would be allowing an array of loader functions to allow for cleaner code when dynamically importing multiple components  & optionally add [Promise.all](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all) support via a new `concurrent` option.

### Example
```tsx
// Synchronous iterable dynamic import
const [ComponentA, ComponentB, ComponentC] = dynamic(
  [
    () => import('./ComponentA'),
    () => import('./ComponentB'),
    () => import('./ComponentC'),
  ],
)
```

```tsx
// Asynchronous iterable dynamic import
const [ComponentA, ComponentB, ComponentC] = dynamic(
  [
    () => import('./ComponentA'),
    () => import('./ComponentB'),
    () => import('./ComponentC'),
  ],
  {
    parallel: true,
  }
)
```

```tsx
// Alternative parameter structure
const [ComponentA, ComponentB, ComponentC] = dynamic(
  () => [
    import('./ComponentA'),
    import('./ComponentB'),
    import('./ComponentC'),
  ],
)
```

This is of course just an idea & a rough idea of what it could look like. Does anyone else have an opinions on this idea? Is it a good idea? Bad idea? Better or a way that already exists that I am not aware of?

**Comment by icyJoseph** [comment]
How would you specify loading components for each? or turn off `ssr`? 

How should this behave if one fails to dynamically load?

**Comment by NuroDev** [comment]
> How would you specify loading components for each? or turn off ssr?

In my mind this would be used for using multiple components of the same time. EG: All components loaded do not need to be ssr, and as such having the single options object as the second parameter applies those options to all provided loaders

>How should this behave if one fails to dynamically load?

This is a bit of a more tricky one to answer. We could return `null` & rely on the dev implementation to add a `.catch` or `try {...} catch {...}` to handle the errors. Alternatively handle it the same way [`Promise.all` currently handles it](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all#rejection) by instead rejecting with the value of the rejection?
If anyone has any better ways / ideas of how to handle that, feel free to throw them out.
